core/tasks.rs

```diff
@@ -0,0 +1,98 @@
+use std::marker::PhantomData;
+use std::ops::DerefMut;
+use std::sync::atomic::AtomicBool;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::task::Context;
+use std::task::Poll;
+
+use futures::lock::MutexGuard;
+use futures::task::AtomicWaker;
+
+type UnsendTask = Box<dyn FnOnce(&mut v8::HandleScope) + 'static>;
+type SendTask = Box<dyn FnOnce(&mut v8::HandleScope) + Send + 'static>;
+
+#[derive(Default)]
+pub(crate) struct V8TaskSpawnerFactory {
+  // TODO(mmastrac): ideally we wouldn't box if we could use arena allocation and a max submission size
+  tasks: Mutex<Vec<SendTask>>,
+  has_tasks: AtomicBool,
+  waker: AtomicWaker,
+}
+
+impl V8TaskSpawnerFactory {
+  pub fn new_same_thread_spawner(self: Arc<Self>) -> V8TaskSpawner {
+    V8TaskSpawner {
+      tasks: self,
+      _unsend_marker: PhantomData,
+    }
+  }
+
+  pub fn new_cross_thread_spawner(self: Arc<Self>) -> V8CrossThreadTaskSpawner {
+    V8CrossThreadTaskSpawner { tasks: self }
+  }
+
+  pub fn has_pending_tasks(&self) -> bool {
+    self.has_tasks.load(std::sync::atomic::Ordering::SeqCst)
+  }
+
+  pub fn poll_inner(&self, cx: &mut Context) -> Poll<Vec<UnsendTask>> {
+    if !self
+      .has_tasks
+      .swap(false, std::sync::atomic::Ordering::SeqCst)
+    {
+      self.waker.register(cx.waker());
+      return Poll::Pending;
+    }
+    let tasks = std::mem::take(self.tasks.lock().unwrap().deref_mut());
+    // SAFETY: we are removing !Send bounds as we return the tasks here
+    let tasks = unsafe { std::mem::transmute(tasks) };
+    Poll::Ready(tasks)
+  }
+
+  fn spawn(&self, task: Box<dyn FnOnce(&mut v8::HandleScope) + Send>) {
+    self.tasks.lock().unwrap().push(task);
+    // TODO(mmastrac): can we use a looser ordering here?
+    self
+      .has_tasks
+      .store(true, std::sync::atomic::Ordering::SeqCst);
+    self.waker.wake();
+  }
+}
+
+/// Allows for submission of v8 tasks on the same thread.
+#[derive(Clone)]
+pub struct V8TaskSpawner {
+  // TODO(mmastrac): can we split the waker into a send and !send one?
+  tasks: Arc<V8TaskSpawnerFactory>,
+  _unsend_marker: PhantomData<MutexGuard<'static, ()>>,
+}
+
+impl V8TaskSpawner {
+  pub fn spawn<F>(&self, f: F)
+  where
+    F: FnOnce(&mut v8::HandleScope) + 'static,
+  {
+    // SAFETY: we are transmuting Send into a !Send handle but we can guarantee this object will never
+    // leave the current thread because `V8TaskSpawner` is !Send.
+    let task: Box<dyn FnOnce(&mut v8::HandleScope<'_>)> = Box::new(f);
+    let task: Box<dyn FnOnce(&mut v8::HandleScope<'_>) + Send> =
+      unsafe { std::mem::transmute(task) };
+    self.tasks.spawn(task)
+  }
+}
+
+/// Allows for submission of v8 tasks on any thread.
+#[derive(Clone)]
+pub struct V8CrossThreadTaskSpawner {
```

Comment:
After thinking on this a bit, I don't think FFI callbacks can use the same-thread `V8TaskSpawner` at all: Same thread callbacks are always necessarily responses to some JS calls going into native code and that native code then synchronously calling an FFI callback C function pointer and expecting a synchronous return value. So that just needs to be a sort of manually handled call with a `CallbackScope` etc.

I wonder if the same-thread task spawner even makes much sense. From a callback perspective at least it doesn't seem to make much sense as it cannot be used to do synchronous callbacks which is the only thing you want to do when on the same thread.

Reply:
There's a few cases where the same-thread spawner works, but it's mainly for newer functionality like @littledivy's WASM rewrite where we need to call the `finish` method from a tokio task that doesn't have a scope associated with it.

Reply:
That sounds like a damn nice use case then! Really nice <3

Reply:
Thinking on it more, I suspect the same-thread task spawner wouldn't be useful for FFI as it would be better to just go right into the callback function, but it does make me wonder if we should consider a specific "v8 callback" API in the future.